core/psh/psh.c

```diff
@@ -972,7 +975,113 @@ static void psh_signalstop(int sig)
 }
 
 
-static int psh_run(void)
+static int psh_logincredentget(char *buff, int ispasswd, int maxinlen)
+{
+	int inputlen = -1;
+	char c;
+
+	for (;;) {
+		if (inputlen < 0) {
+			if (ispasswd)
+				fprintf(stdout, "Password: ");
+			else
+				fprintf(stdout, "Login: ");
+
+			fflush(stdout);
+			inputlen++;
+		}
+
+		if (read(STDIN_FILENO, &c, 1) != 1)
+			continue;
+			
+		if (c == '\n') {
+			if (inputlen <= maxinlen && inputlen > 0) {
+				buff[inputlen] = '\0';
+				return --inputlen;
+			}
+			else {
+				inputlen = -1;
+				if (ispasswd) {
+					memset(buff, '\0', sizeof(char)*maxinlen);
+					return 0;
+				}
+				else
+					continue;
+			}
+		}
+		else if (isprint(c)) {
+			if (inputlen < maxinlen)
+				buff[inputlen] = c;
+			inputlen++;
+			continue;
+		}
+	}
+	return -EINTR;
+}
+
+
+static int psh_login(void)
+{
+	int err = 0, auth = -EACCES;
+	const int maxlen = 32;
+	char username[maxlen+1], passwd[maxlen+1], *shadow;
+	struct passwd *userdata;
+	struct termios origattr, ttyattr;
+
+	/* get/prepare tty settings */
+	if ((err = tcgetattr(STDIN_FILENO, &origattr)) < 0 || (err = tcgetattr(STDIN_FILENO, &ttyattr)) < 0) {
+		sleep(1);
+		fprintf(stderr, "psh: failed to save terminal settings\n");
+		return -ENOTTY;
+	}
+	if ((err = isatty(STDOUT_FILENO)) == 0) {
```

Comment:
I guess the order should be reversed, first checking if it is a terminal and then reading the property:
```
isatty()
tcgetattr()
```

Reply:
Yes, I also noticed that. After moving tty attribute setters and getters to subroutine the order is correct